### ISSUE TITLE
Compatibility with django-polymorphic

### ISSUE THREAD

**Issue by Yurimord** [theatlantic/django-nested-admin#86]
I want to use polymorphic form inside  your nested admin
I got error
  File "/usr/local/lib/python2.7/dist-packages/nested_admin/nested.py", line 48, in __iter__
    for inline_admin_form in super(NestedInlineAdminFormset, self).__iter__():
  File "/usr/local/lib/python2.7/dist-packages/django/contrib/admin/helpers.py", line 273, in __iter__
    self.formset, self.formset.empty_form,
  File "/usr/local/lib/python2.7/dist-packages/polymorphic/formsets/models.py", line 247, in empty_form
    raise RuntimeError("'empty_form' is not used in polymorphic formsets, use 'empty_forms' instead.")
RuntimeError: 'empty_form' is not used in polymorphic formsets, use 'empty_forms' instead.

Each app works  good separately.
Can you help to figure out how to fix it?
Is anybody using  this package together? 
Thanks

**Comment by joetsoi** [comment]
Just in case anyone else needs this in the future. You need to add this snippet to code and have your admin class derive from it.
```
class PolymorphicNestedModelAdmin(NestedModelAdmin, PolymorphicInlineSupportMixin):
    def get_inline_formsets(self, request, formsets, inline_instances,
                            obj=None, allow_nested=False):
        """Admin that fixes compatibility between nested and polymorphic admins

        Calls super on NestedModelAdmin and passes the returned list to a copy of
        PolymorphicInlineSupportMixin.get_inline_formsets()

        Both get_inline_formsets() on the parent classes need to be called in order
        for both to work instead of just one in the mro case.
        """
        inline_admin_formsets = super(NestedModelAdmin, self).get_inline_formsets(
            request, formsets, inline_instances, obj, allow_nested
        )
        # this is a copy of PolymorphicInlineSupportMixin.get_inline_formsets
        for inline_admin_formset in inline_admin_formsets:
            if isinstance(inline_admin_formset.formset, BasePolymorphicModelFormSet):
                # This is a polymorphic formset, which belongs to our inline.
                # Downcast the admin wrapper that generates the form fields.
                inline_admin_formset.__class__ = PolymorphicInlineAdminFormSet
                inline_admin_formset.request = request
                inline_admin_formset.obj = obj
        return inline_admin_formsets
```

Both `get_inline_formsets` need to be called on `NestedInlineAdminFormset` and `PolymorphicInlineSupportMixin`, due to mro, only one gets called, so adding this class works around that.

**Comment by fdintino** [comment]
@atleta When that comment was posted, I was pretty sure that this solution would not work, but someone else gave it a thumbs up so I didn't comment. Both django-nested-admin and django-polymorphic extend the html and javascript of inlines in significant ways, and they are fundamentally incompatible with each other. Adding polymorphic support for django-nested-admin would require pretty significant changes, which is why I haven't tried tackling this. If there is a lot of demand for this feature then I could look at what can be done—I think I'd be in a good position to do that since I've also contributed to django-polymorphic in the past.

**Comment by brownish** [comment]
@fdintino I would absolutely love for this to be implemented!

**Comment by fdintino** [comment]
It is in the works! Will probably ship soon.

**Comment by brownish** [comment]
Fantastic! Thanks so much!

**Comment by brownish** [comment]
@fdintino I've been trying to get this to work and I'm not sure if I'm just not remembering how to do it correctly or if it's a bug I'm encountering.

Here is the basics of my admin.py With this, I'm able to select my first level inline and the form shows up correctly except I don't have the next child inline (Pottery Decoration inline). Am I doing this right?


```python
class PotteryDecorationInline(StackedPolymorphicInline, nested_admin.NestedStackedInlineMixin):
    class ImpressedInline(StackedPolymorphicInline.Child, nested_admin.NestedStackedInlineMixin):
        model = ImpressedDecoration

    class StampedInline(StackedPolymorphicInline.Child, nested_admin.NestedStackedInlineMixin):
        model = StampedDecoration

    model = PotteryDecoration
    child_inlines = (
        ImpressedInline,
        StampedInline
    )
    list_display = ('__class__', 'default')


class PotteryMorphologyInline(StackedPolymorphicInline, nested_admin.NestedStackedInlineMixin):
    class BodyInline(StackedPolymorphicInline.Child, nested_admin.NestedStackedInlineMixin):
        model = PotteryBody
        inlines = (PotteryDecorationInline,)

    class ShoulderInline(StackedPolymorphicInline.Child, nested_admin.NestedStackedInline):
        model = PotteryShoulder
        inlines = (PotteryDecorationInline,)

    class RimInline(StackedPolymorphicInline.Child, nested_admin.NestedStackedInlineMixin):
        model = PotteryRim
        inlines = (PotteryDecorationInline, )

    model = PotteryMorphology
    child_inlines = (
        BodyInline,
        ShoulderInline,
        RimInline,
    )
    list_display = ('__class__', 'default')


@admin.register(Pottery)
class PotteryAdmin(PolymorphicInlineSupportMixin, nested_admin.NestedModelAdmin, ArtifactAdmin):
    base_model = Pottery
    fieldsets = ArtifactAdmin.base_fieldsets + (
        ('Pottery-Specific Info', {
            'fields': ('material', 'thickness', 'temper'),
            'classes': ('grp-collapse grp-open',),
        }),
    )
    inlines = (PotteryMorphologyInline,)


@admin.register(Artifact)
class ArtifactParentAdmin(PolymorphicParentModelAdmin, nested_admin.NestedModelAdmin):
    base_model = Artifact
    child_models = (Lithic, Pottery, Zoological, Botanical, Matrix, Historic, Mineral)
    list_display = ('accession_year', 'accession_number', 'artifact_class')
```

**Comment by fdintino** [comment]
@brownish django-polymorphic support is available in the 3.2.0 release (though it should still be considered in beta). There is an example of how to use the integration [in the latest docs](https://django-nested-admin.readthedocs.io/en/latest/integrations.html).

Besides following the example there, one other thing you'll need to do is to replace:

```python
class ArtifactParentAdmin(PolymorphicParentModelAdmin,
                          nested_admin.NestedModelAdmin):
```

with

```python
class ArtifactParentAdmin(nested_admin.NestedModelAdminMixin,
                          PolymorphicParentModelAdmin):
```

(note the order being reversed, and using `NestedModelAdminMixin` rather than `NestedModelAdmin`)

**Comment by brownish** [comment]
Thanks, I'll give it a try.
